test/Feature/UpsertSpec.hs

```diff
@@ -0,0 +1,183 @@
+        it "INSERTs and UPDATEs row on composite pk conflict" $
+          request methodPost "/employees" [("Prefer", "return=representation"), ("Prefer", "resolution=merge-duplicates")]
+            [json| [
+              { "first_name": "Frances M.", "last_name": "Roe", "salary": "30000" },
+              { "first_name": "Peter S.", "last_name": "Yang", "salary": 42000 }
+            ]|] `shouldRespondWith` [json| [
+              { "first_name": "Frances M.", "last_name": "Roe", "salary": "$30,000.00", "company": "One-Up Realty", "occupation": "Author" },
+              { "first_name": "Peter S.", "last_name": "Yang", "salary": "$42,000.00", "company": null, "occupation": null }
+            ]|]
+            { matchStatus = 201
+            , matchHeaders = [matchContentTypeJson]
+            }
+
+      context "when Prefer: resolution=ignore-duplicates is specified" $ do
+        it "INSERTs and ignores rows on pk conflict" $
+          request methodPost "/tiobe_pls" [("Prefer", "return=representation"), ("Prefer", "resolution=ignore-duplicates")]
+            [json|[
+              { "name": "PHP", "rank": 9 },
+              { "name": "Python", "rank": 10 }
+            ]|] `shouldRespondWith` [json|[
+              { "name": "PHP", "rank": 9 }
+            ]|]
+            { matchStatus = 201
+            , matchHeaders = [matchContentTypeJson]
+            }
+
+        it "INSERTs and ignores rows on composite pk conflict" $
+          request methodPost "/employees" [("Prefer", "return=representation"), ("Prefer", "resolution=ignore-duplicates")]
+            [json|[
+              { "first_name": "Daniel B.", "last_name": "Lyon", "salary": "72000", "company": null, "occupation": null },
+              { "first_name": "Sara M.", "last_name": "Torpey", "salary": 60000, "company": "Burstein-Applebee", "occupation": "Soil scientist" }
+            ]|] `shouldRespondWith` [json|[
+              { "first_name": "Sara M.", "last_name": "Torpey", "salary": "$60,000.00", "company": "Burstein-Applebee", "occupation": "Soil scientist" }
+            ]|]
+            { matchStatus = 201
+            , matchHeaders = [matchContentTypeJson]
+            }
+
+      it "succeeds if the table has only PK cols and no other cols" $ do
+        request methodPost "/only_pk" [("Prefer", "return=representation"), ("Prefer", "resolution=ignore-duplicates")]
+          [json|[ { "id": 1 }, { "id": 2 }, { "id": 3} ]|]
+          `shouldRespondWith`
+          [json|[ { "id": 3} ]|] { matchStatus = 201 , matchHeaders = [matchContentTypeJson] }
+        request methodPost "/only_pk" [("Prefer", "return=representation"), ("Prefer", "resolution=merge-duplicates")]
+          [json|[ { "id": 1 }, { "id": 2 }, { "id": 4} ]|]
+          `shouldRespondWith`
+          [json|[ { "id": 1 }, { "id": 2 }, { "id": 4} ]|] { matchStatus = 201 , matchHeaders = [matchContentTypeJson] }
+
+      it "succeeds and ignores the Prefer: resolution header if the table has no PK" $
```

Comment:
What about this alternative: if a table has no pk then make the Prefer header `resolution=ignore-duplicates` translate to `ON CONFLICT DO NOTHING` (which has an empty conflict target). But have `resolution=merge-duplicates` raise an error on tables without a pk because there's no obvious conflict target other than the primary keys. Currently it appears that the entire `ON CONFLICT` clause is dropped for these tables.

I feel like if a client explicitly requests merge-duplicates then it's an error to ignore it on tables without a pk. Is there a use-case you have in mind that I am overlooking?

Reply:
I chose to drop the `ON CONFLICT` clause for table with no pks because I thought that `Prefer` semantics meant the condition it's optional and shouldn't give an error, also according to [RFC 7240](https://tools.ietf.org/html/rfc7240):
> A server that does not recognize or is unable to comply with
   particular preference tokens in the Prefer header field of a request
   MUST ignore those tokens and continue processing instead of signaling
   an error.

I thought to later add the `Preference-Applied` feature so the client can check the success of the condition.

Reply:
Great point that this is only a Prefer header so we shouldn't raise an error if the header can't be applied.

About tables with no primary keys, we could conceivably do `ON CONFLICT` with no columns in parentheses. The docs say:

> If no list of column names is given at all, the default is all the columns of the table in their declared order

Only problem is this might cause an expensive sequential scan for duplicates rather than using indices, so it would put extra load on the server. Your choice to drop the ON CONFLICT for tables without keys is probably the right thing to do.

Would it be difficult to include a `Preference-Applied` response header for resolution? Might help people debug a surprising result.